jmix-flowui/flowui/src/main/java/io/jmix/flowui/component/navigationmenubar/NavigationMenuBar.java

```diff
@@ -0,0 +1,830 @@
+        Preconditions.checkNotNullArgument(menuItem);
+
+        addMenuItemRecursive(menuItem, null, null);
+
+        rootMenuItems.add(menuItem);
+    }
+
+    protected void addMenuItemRecursive(AbstractMenuItem<?> childMenuItem,
+                                        @Nullable ParentMenuItem parentMenuItem,
+                                        @Nullable Integer index) {
+        checkItemIdDuplicate(getMenuItemIdNN(childMenuItem));
+
+        JmixSubMenu parentSubMenu = getParentSubMenu(parentMenuItem);
+
+        if (childMenuItem.isMenu()) {
+            addMenuItemToSubMenu(childMenuItem, parentSubMenu, index);
+
+            ParentMenuItem childParentMenuItem = (ParentMenuItem) childMenuItem;
+            for (AbstractMenuItem<?> item : childParentMenuItem.getChildItems()) {
+                addMenuItemRecursive(item, childParentMenuItem, null);
+            }
+        } else if (childMenuItem.isSeparator()) {
+            if (parentSubMenu != null) {
+                addComponentToSubMenu(childMenuItem, parentSubMenu, index);
+            } else {
+                throw new UnsupportedOperationException("Separators are not supported on the root level");
+            }
+        } else {
+            addMenuItemToSubMenu(childMenuItem, parentSubMenu, index);
+        }
+
+        attachMenuItem(childMenuItem);
+    }
+
+    protected String getMenuItemIdNN(AbstractMenuItem<?> menuItem) {
+        return menuItem.getId().orElseThrow(() -> new IllegalStateException("Menu item id is not defined"));
+    }
+
+    protected void checkItemIdDuplicate(String id) {
+        if (allMenuItems.containsKey(id)) {
+            throw new IllegalArgumentException("Menu item with id \"%s\" already exists".formatted(id));
+        }
+    }
+
+    @Nullable
+    protected JmixSubMenu getParentSubMenu(@Nullable ParentMenuItem parentMenuItem) {
+        if (parentMenuItem == null) {
+            return null;
+        }
+
+        JmixMenuItem itemWrapper = parentMenuItem.getMenuItemWrapper();
+        if (itemWrapper == null) {
+            throw new IllegalArgumentException("Parent menu item is not attached to the menu");
+        }
+        return itemWrapper.getSubMenu();
+    }
+
+    protected void addMenuItemToSubMenu(AbstractMenuItem<?> menuItem,
+                                        @Nullable JmixSubMenu subMenu,
+                                        @Nullable Integer index) {
+        HasMenuItemsEnhanced hasMenuItems = subMenu != null ? subMenu : getContent();
+
+        JmixMenuItem wrapper = index == null
+                ? hasMenuItems.addItem(menuItem)
+                : hasMenuItems.addItemAtIndex(index, menuItem);
+        menuItem.setMenuItemWrapper(wrapper);
+
+        if (subMenu == null) {
+            menuItem.addClassNames(ROOT_MENU_ITEM_CLASS_NAME);
+        }
+    }
+
+    protected void addComponentToSubMenu(Component component, JmixSubMenu subMenu, @Nullable Integer index) {
```

Comment:
Let's avoid using classes wrappers for primitive types. `null` can be replaced with `-1` or list's size, e.g. as this is done for actions:

```
default void addAction(Action action) {
    addAction(action, getActions().size());
}
```